keps/sig-storage/2857-runtime-assisted-pv-mounts/README.md

```diff
@@ -0,0 +1,1387 @@
+location specified in the pod spec.
+
+While the pod runs:
+- The node CSI plugin receives a CSI `NodeGetVolumeStats` call. In response, the CSI
+plugin passes the shared fs export path (corresponding to the volume) to the
+Kubelet. Based on the response, the Kubelet invokes `RuntimeGetFileSystemStats`
+on the microvm runtime's unix domain socket for handling CSI Runtime APIs.
+
+When the pod terminates:
+- The runtime dismounts the filesystem in the sandbox environment
+- The node CSI plugin receives CSI `NodeVolumeUnpublish` and cleanups any state
+but skips dismounting the file system on the PV.
+
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+Deferring filesystem mounts to the runtime may require restricted PVC access
+modes. Specifically, ReadWriteOncePod will need to be specified on PVCs to
+prevent multiple pods attempting to mount the same PVC when scheduled on the
+same node if the PVC is bound to a block based PV with XFS or ext4 file system.
+Such restrictions, however, are not necessary if the filesystem to be mounted
+can support parallel mounts.
+
+In the context of microvm runtimes, using PVCs specifying access modes that allow
+multiple writers like: ReadWriteOnce (which allows multiple pods in the same
+node to mount a PV in read/write) and ReadWriteMany should be used with
+caution depending on the isolation goals. Multiple writer access modes (within
+the same or distinct nodes) allow a compromised/malicious pod to write data
+into a shared volume that may affect: [1] other pods reading the data through
+corrupt/fuzzed data and [2] other writers through denial of service attacks like
+inode exhaustion on the volume. As a result, goals around isolation between pods
+may be compromised. A cluster admin can configure a webhook/OPA policies to
+restrict the set of access modes that can be specified on PVCs that are referred
+by pods associated with a microvm runtime.
+
+A pod typically maps to the isolated sandbox environment in the context of
+microvm runtimes. Individual containers in the pod within the sandbox are not
+expected to be isolated with the same guarantees that exist across pods. To
+align with these isolation goals, restrictions around the ability of multiple
+containers within a pod mounting the same PV should not be necessary and
+considered beyond the scope of this KEP.
+
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+
+## Design Details
```

Comment:
As you can see, each feature requires it's own workaround. This will become inevitably very fragile, as we will add more features to CSI and kubelet. It will be hard to make sure various CSI/kubelet versions work well with VM versions, which can be older or event newer. Luckily, we have quite good test coverage, so we can test these combinations easily, still, IMO it will be a huge effort to make sure that everything really works and keep it working.

One additional feature that comes to my mind: how will you deal with a pod with multiple containers, sharing the same volume? Will each container have its own VM? Then the volume needs to be mounted only once, probably on the host.

Reply:
Agree that there will be limitations and extensive test coverage will be necessary.

Regarding multiple containers sharing the same volume in a pod: the sandbox for the VM based runtimes (like Kata) is at the pod level. Within the sandbox VM environment, it's just regular containers bind mounting the same volume to share them.

Reply:
Updated the KEP to account for handling all the above scenarios through a direct handshake between CSI plugin and a container runtiem.